crypto/modes/gcm128.c

```diff
@@ -669,33 +669,37 @@ void gcm_gmult_4bit_x86(u64 Xi[2], const u128 Htable[16]);
 void gcm_ghash_4bit_x86(u64 Xi[2], const u128 Htable[16], const u8 *inp,
                         size_t len);
 #  endif
-# elif defined(__arm__) || defined(__arm) || defined(__aarch64__)
+# elif (defined(__arm__) || defined(__arm) || defined(__aarch64__)) && defined(GHASH_ASM)
 #  include "arm_arch.h"
 #  if __ARM_MAX_ARCH__>=7
 #   define GHASH_ASM_ARM
 #   define GCM_FUNCREF_4BIT
-#   define PMULL_CAPABLE        (OPENSSL_armcap_P & ARMV8_PMULL)
 #   if defined(__arm__) || defined(__arm)
 #    define NEON_CAPABLE        (OPENSSL_armcap_P & ARMV7_NEON)
 #   endif
 void gcm_init_neon(u128 Htable[16], const u64 Xi[2]);
 void gcm_gmult_neon(u64 Xi[2], const u128 Htable[16]);
 void gcm_ghash_neon(u64 Xi[2], const u128 Htable[16], const u8 *inp,
                     size_t len);
+#   if __ARM_MAX_ARCH__>=8
+#    define PMULL_CAPABLE        (OPENSSL_armcap_P & ARMV8_PMULL)
 void gcm_init_v8(u128 Htable[16], const u64 Xi[2]);
 void gcm_gmult_v8(u64 Xi[2], const u128 Htable[16]);
 void gcm_ghash_v8(u64 Xi[2], const u128 Htable[16], const u8 *inp,
                   size_t len);
+#   endif
 #  endif
 # elif defined(__sparc__) || defined(__sparc)
 #  include "sparc_arch.h"
-#  define GHASH_ASM_SPARC
-#  define GCM_FUNCREF_4BIT
+#  if defined(__arch64__)
+#   define GHASH_ASM_SPARC
+#   define GCM_FUNCREF_4BIT
```

Comment:
Again, just because it's counter-intuitive, it doesn't mean it's wrong [or unintentional]. Though reason is a bit more subtle than that in ARM case. Just like in ARM case, 32-bit application can take advantage of the module in question, **vendor** compiler unfortunately doesn't pre-define any additional macros if you ask for 32-bit code that can execute on V9 processor. As result, you're stuck with `__sparc` macro as the only option. On additional note, Linux also implements dual-ABI setup, so that throwing in `__arch64__` cripples 32-bit applications. In other words if this actually solves some problem somewhere, it creates problem elsewhere.

Reply:
Thanks for the explanation, I will revert this.